addons/sale/models/sale_order.py

```diff
@@ -788,6 +788,15 @@ def _unlink_except_draft_or_cancel(self):
                     "You can not delete a sent quotation or a confirmed sales order."
                     " You must first cancel it."))
 
+    def write(self, vals):
+        new_partner_id = vals.get('partner_id')
+        if new_partner_id:
+            for order in self:
+                old_partner_id = order.partner_id.id
+                if order.state in ('sent', 'sale') and old_partner_id != new_partner_id:
```

Comment:
You're just checking that the customer is not the same as before, but he could already be a follower. You're therefore filtering some cases, not all. So I'd remove the check on the partner and call `message_subscribe` as soon as `partner_id` is in vals.

Reply:
Indeed, message_subscribe checks whether to provided partners are already subscribed, so the check is not strictly necessary. I removed the check and simplified the code a bit.